📝 Performance doc (#736)
sequinstream/sequin: docs/performance.mdx

---
title: "Performance"
description: "Sequin change data capture (CDC) performance benchmarks"
icon: "gauge-high"
iconType: "solid"
---

Sequin delivers industry-leading performance for change data capture (CDC), consistently achieving sub-200ms latency at 2,000 sustained operations per second. Here's how Sequin compares to other solutions:

| Tool | Latency at 2k ops/s |
|------|-------------------|
| **Sequin** | **192ms** |
| Debezium | 234ms |
| PeerDB | 30 seconds |
| Fivetran | 5 minutes |
| Airbyte | 1+ hours |

> ## Test Setup

Reviewer: ```suggestion
## Test setup
```

The load generation script applies a mixed workload of `INSERT`, `UPDATE`, and `DELETE` operations to the `benchmark_records` table at a total throughput of >= 2,000 operations per second.

The Posgres schema includes a single table with this schema:
```
                                           Table "public.benchmark_records"
     Column      |            Type             | Collation | Nullable |                    Default
-----------------+-----------------------------+-----------+----------+-----------------------------------------------
 id              | integer                     |           | not null | nextval('benchmark_records_id_seq'::regclass)
 string_field    | text                        |           |          |
 numeric_field   | numeric                     |           |          |
 timestamp_field | timestamp with time zone    |           |          |
 json_field      | jsonb                       |           |          |
 inserted_at     | timestamp without time zone |           |          | now()
 updated_at      | timestamp without time zone |           |          | now()
```

> Records delivered to Kafka are annotated with a `delivered_at` timestamp. Full end-to-end latency is calculated by subtracting the `delivered_at` timestamp from the `updated_at` timestamp.

Reviewer: I assume `updated_at` is set by the script on updates?

Author: correct

> ### Throughput Stability

Reviewer: ```suggestion
### Throughput stability
```

Sequin maintains consistent performance even under sustained load: 

```bash
Average throughput: 2,054.4 ops/sec
Average latency: 192ms
99th percentile latency: 245ms
```

By comparison, Debezium achieves a slightly higher latency of 234ms at similar throughput:

```bash
Average throughput: 2000.6 ops/sec
Average latency: 234ms
99th percentile latency: 286ms
```

> ### Resource Utilization

Reviewer: ```suggestion
### Resource utilization
```

> The Sequin ECS task makes efficient use of provisioned resources, employing parallelism with lightweight processes. During the test, the Sequin ECS task averaged 34% CPU utilization and 10% memory utilization.

Reviewer: ```suggestion
Sequin was deployed to EC2 via Elastic Container Service (ECS). During the test, the Sequin ECS task averaged 34% CPU utilization and 10% memory utilization.
```

## Next Steps

> Ready to see Sequin's performance for yourself?

Reviewer: You could maybe even use this component:

```
<CardGroup cols={2}>
  <Card title="Guide: Connect Postgres" icon="elephant" href="/connect-postgres">
    Connect your Postgres database to Sequin.
  </Card>
  <Card title="Guide: Setting up an SQS sink" icon="aws" href="/how-to/stream-postgres-to-sqs">
    Setup an SQS sink to send changes to your queue.
  </Card>
</CardGroup>
```

> - [Quick Start Guide](/quickstart)

Reviewer: ```suggestion
- Check out our [quickstart with Kafka](/quickstart/kafka)
```

> - [Deploy to Production](/how-to/deploy-to-production)

Reviewer: ```suggestion
- See our how-to on [deploying to production](/how-to/deploy-to-production)
```

> - [Compare to Other CDC Tools](/compare)

Reviewer: ```suggestion
- [Compare Sequin to other CDC tools](/compare)
```
